### ISSUE TITLE
Components: Resolving unintended key events in IME composition

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#45605]
Related to: #45510, #45599

This tracking issue seeks to resolve the problem of events occurring in IME composition mode causing unintended behavior in some components.

## What is IME composition ?

One of the input systems in which the IME composition are used is Japanese.

Because of the vast number of characters in the Japanese language ( including Kanji), it is not possible to input characters directly on a standard keyboard. Therefore, after entering "hiragana" characters, we choose which character to convert to and confirm with the Enter key.

- When you begin typing a character, an underline appears indicating that the expected character has not been finalized.
- Use the `space` or `up/down` keys to move through the list of candidates.
- Use the 'left' and 'right' keys to move the cursor position.
- Press the `delete` or `backspace` key to delete characters before or after the current cursor position
- Finally, press the `enter` key to confirm your input.
- Press the `escape` key to hide input candidates or cancel what you are typing.

<details>
<summary>Screencast</summary>

https://user-images.githubusercontent.com/54422211/200585845-c21654bd-7af1-4b56-be9e-4a9db8eab255.mp4

</details>

## What is the problem if it is IME composition ?

The keyboard operations described above are only for finalizing characters. However, for some components, these keystrokes are monitored and may cause unintended behavior.

## What is the cause and solution?

I found the following differences about key events when character input is not finalized.

- `event.code`: returns the pressed key itself
- `event.keyCode`: returns the immediately preceding key
- `event.key`: returns the string Process

Therefore, in components that accept text input and use `event.code`, I believe it is appropriate to rewrite them using `event.key`.

@edit: It turns out that simply replacing `event.code` with `event.key` is insufficient. As noted in [this comment](https://github.com/WordPress/gutenberg/pull/45607#discussion_r1018323807), it would be desirable to add checks for `event.isComposing` and `event.keyCode`.

## Components to be fixed

After a series of tests, I believe that the following components are causing problems and should be fixed:

- [x] `FormTokenField`: Duplicate tags are entered
  - #45607
- [x] `ComboboxControl`: When the `enter` key is pressed, both candidate characters and typed characters may be inserted.
  - #46827
- [x] `Modal`: Pressing the `escape` key to cancel the input state causes the modal to disappear.
  - #46453
- [x] `AutocompleteUI`: Forward slash inserter doesn't work in Japanese
  - #45510
  - #45626
- [x] Follow-up fix to `TokenInput` component (see #46827 ' s description)

The following components may have a lower priority because I believe that there are few cases where text input elements will be placed.

- [x] `ConfirmDialog`: Pressing the `escape` key to cancel the input state causes the modal to disappear.
  - #46453
- [x] `Guide`: Pressing the `escape` key to cancel the input state causes the modal to disappear.
  - #46453

**Comment by t-hamano** [comment]
@ciampo @mirka 
If there are other issues that need to be addressed, please advise me 🙏

**Comment by t-hamano** [comment]
Update: The `ConfirmDialog` and `Guide` components in the task list use `Modal` components internally, so this was automatically fixed by #46453. I temporarily added input elements to the two components, and IME input works correctly.

https://github.com/WordPress/gutenberg/assets/54422211/e035d648-ff03-40de-acb1-6b1843d1e249

**Comment by t-hamano** [comment]
Once #60090 is merged, all IME composition issues in the component package should be resolved. Therefore, I would like to close this issue.
